server/app/jobs/certificate_renew_job.rb

```diff
@@ -0,0 +1,91 @@
+class CertificateRenewJob
+  include Celluloid
+  include Logging
+  include CurrentLeader
+  include WaitHelper
+
+  RENEW_INTERVAL = 5 * 60 # Run the renew check every 5 mins
+
+  def initialize(perform = true)
+    async.perform if perform
+  end
+
+  def perform
+    # TODO Maybe sleep a while first to allow everything else to start up before running this loop?
+    info 'starting to watch certificate renewals'
+    every(RENEW_INTERVAL) do
+      if leader?
```

Comment:
In theory this has a possibility for a race. As the renewal process is bit slow and if one has many certs, leader can change while processing all certs. That would mean that the new leader might process same certs and thus make "double-renewal"

Reply:
Maybe better to use `with_dlock` instead of `leader?` for this. It doesn't necessarily have to be the leader that handles it?

Although in that case, the job would need to scale/randomize the `every` loop to spread the intervals across all of the workers, or it would happen every `RENEW_INTERVAL`/number-of-workers...

IMO this doesn't need to happen as frequently as once every 5 minutes... perhaps only once a day... there just needs to be some room for retrying.

Reply:
It doesn't matter whether it is the leader or not, just needs to be ONE worker that handles it. Thus I thought it would be most simple to handle this only on leader. IMO running this only on leader makes it super simple and the possible race should be somewhat theoretical. Putting in loop randomization and dlocks makes it unnecessarily complex IMO.

Reply:
Yes, the renewal loop should be ok to run once a day. Now it starts renewals 7 days before expiration, so it should have at least 6 times to retry.

Reply:
> Putting in loop randomization and dlocks makes it unnecessarily complex IMO.

Yes, the dlocks are not really that suitable for trying to schedule tasks across all the workers.

> IMO running this only on leader makes it super simple and the possible race should be somewhat theoretical.

Hmm... the leader period is ~60s, I think it's entirely normal to see the leader change as often as once a minute?

```
I, [2017-09-15T10:15:53.434178 #8]  INFO -- LeaderElectorJob: participating leader elections
I, [2017-09-15T10:15:53.456462 #8]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:15:53.511674 #18]  INFO -- LeaderElectorJob: participating leader elections
I, [2017-09-15T10:15:54.078487 #16]  INFO -- LeaderElectorJob: participating leader elections
I, [2017-09-15T10:15:54.502236 #12]  INFO -- LeaderElectorJob: participating leader elections
I, [2017-09-15T10:16:51.483532 #16]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:17:49.505809 #18]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:18:47.527563 #8]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:19:45.553094 #18]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:20:43.578673 #16]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:21:41.591414 #18]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:22:39.621962 #12]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:23:37.643350 #16]  INFO -- LeaderElectorJob: won election ♚
I, [2017-09-15T10:25:33.672318 #12]  INFO -- LeaderElectorJob: won election ♚
```

Reply:
If the different `CertificateRenewJob` every loops start drifting out of phase, then I imagine it's entirely possible to hit a race here. It just needs the `every` loop on the current leader to be the worker where the `every` loop is firing the earliest, and then for a routine leader re-election to happen, which means that the `every` loop might fire again very soon on the new leader.

Maybe keep the `leader?` for the scheduling, but then stick a `with_dlock` around the call to `renew_certificates`?

Reply:
yeah, makes sense

Reply:
done